### ISSUE TITLE
RFD - A11y Testing Phase 1

### ISSUE THREAD

**Issue by gabalafou** [Quansight-Labs/jupyter-a11y-mgmt#97]
| Status       | Accepted ✅ |
|--------------|----------------------------------------|
| Author(s)    | @gabalafou |
| Date Created | 12-04-2022 |
| Date Last updated| 21-04-2022 |
| Decision deadline | 19-04-2022 |

# Six week testing marathon 

A word of acknowledgment: many of the thoughts below are not my own. I could not have produced this document without Tony. I have also been aided in discussions with Isabela and Tania. My thinking has also been shaped by online discussions from people in the web a11y community, as well as the [JupyterLab benchmarks repo](https://github.com/jupyterlab/benchmarks). 

## Summary

Refresher—our CZI grant comprises essentially three steps:

1. Test (JupyterLab for accessibility issues)
2. Fix (issues in JupyterLab)
3. Share (learnings with the wider open source community)

What are Tony and I hoping to solve over the next six weeks? We haven't made significant progress yet on the CZI grant deliverables, in particular the first one, testing. We hope that by rallying together **as a team**, we can make significant progress towards our testing deliverable.

## User benefit

This section is mostly a discussion of testing philosophy. It attempts to articulate a justification for testing, upon which the concrete proposal is based. But if you just want to get to the proposal, jump to the next section.

So what is testing trying to solve? Over the past months, I've heard everything from motivating developers to fix accessibility issues by creating squeaky-wheel CI failures to helping developers understand how their code changes affect accessibility.

A similar challenge of scope is that JupyterLab accessibility issues touch upon so many distinct and disparate areas, not only the JupyterLab app, but also JupyterLab docs, extensions, and other critical parts of the ecosystem (viz., the notebook reader experience versus the notebook writer experience). However, the automated testing needs of docs are fairly different from the needs of a stateful, interactive web app like JupyterLab.

So what is testing trying to solve? In an ideal world, I think it solves the issue that Tania has mentioned in previous team meetings: giving us a "control room" view of what's broken, what needs to be fixed—letting us know where we are. 

But I don't think we can actually build that. For one, there's an issue of limited resources. And for another, there's an issue that some of the most important accessibility issues are impossible to automate (for example, basically everything semantic: good link text, logical and meaningful headings, etc.). So, **at most**, we can get a limited control room view of things that the computer thinks need to be fixed.

So I reject this raison d'être for tests. I don't believe that they are there to let us know the state of accessibility. I believe that they serve a purpose that is closer to what other automated tests serve, which in my experience is some combination of helping to document the expected behavior of the app, to catch bugs before they happen, and to prevent bugs from re-occurring.

Where this analogy breaks down is that for general software testing, if you have a very good suite of tests, then you can believe with some confidence that if your tests pass, the application will behave as expected—in theory, anyway. But with accessibility, there is no set of accessibility tests you can run that will give you fairly high confidence that if the tests pass, your application is "accessible," any more than there are a set of security tests you can run against your application and be confident that your application is "secure." As with security, all you can do is use tests to help you avoid common pitfalls. 

So in my view, the main problem that testing is trying to solve is to help us avoid common pitfalls. If we avoid those pitfalls, we decrease the chance that some user will struggle with a particular aspect of our application (for example, color contrast).

This distinction is important because I believe that everything we do should be justified by the criteria of: does this help make and keep the app more accessible? I think automated testing can help us make accessibility gains, but even more powerfully help us to keep the gains we make.

So in my mind, that's the main purpose. For the purposes of our grant, there are some additional purposes that testing can serve:

- Helping to document things we've fixed along the way
- Increasing the visibility of existing pitfalls in JupyterLab
- Providing a potential pathway to contributions from developers who are motivated by CI or by flipping reds to greens on a test report
- Providing a potential pathway to developers who like writing tests
- and more!

## Proposal

Okay, with the why's and wherefore's of the previous section out of the way, now that you understand what overall purpose I think testing serves, in this section I will make a six-week proposal that I think addresses that purpose.

Why six weeks? It's an artificial deadline with not much purpose other than to motivate us. It's also an amount of time that I believe is long enough to produce a significant deliverable. I have divided the proposal into three parts, each with an independent but related deliverable.

Part 1.

A software package to run a suite of accessibility tests against any **app code** PR in JupyterLab (by app code, I mean this does not cover docs yet—visit section below for longer discussion). To begin, our test test suite will include:

- Standard axe-core diagnostics against three to five JupyterLab "views"
- Three to five handwritten machine-as-a-user tests (i.e., Playwright) 

Part 2.

Utilize a service (i.e., GitHub actions) to make it easy for any PR author or reviewer to run our accessibility test suite against their PR. For example, someone could type "Run accessibility tests" in a PR comment and that could trigger a bot to run the tests and report back in a comment on the PR.

Part 3. 
(Beyond six weeks, or six-week stretch goal)

Configure a scheduler to run the test suite on a regular schedule and figure out how and where to publish the results to maximize impact and visibility.

The division of the project into three parts makes it look like these parts are sequential, but a lot of the work should probably be done in parallel.

### Timeline

This is more of an example timeline than a real timeline proposal. This is just to give an idea of how the work might be broken up by time and by team member.

Week 1. (this week). Team alignment (this document).

Week 2. Get axe-core tests into jupyter/accessibility
  - Tony to get Python, Doit, and environment working
  - Tania to help configure CI 
  - Isabela to work with Tony to define 2-4 "pages" to feed to axe-core (keeping color contrast front of mind)
  - Gabriel to pair with Tony to help set up pages for axe-core
 
Week 3. Set up axe-core tests for storage and sharing (GitHub comment on PR, for example)
  - Isabela to select and define 1-2 tests for Playwright + Galata (keeping color contrast front of mind)
  - Gabriel, Tony and Tania to work on wrangling GitHub's CI system
  
Week 4. Add scaffold for Galata + Playwright tests to jupyter/accessibility
  - Isabela to select and define 1-2 more tests
  - Gabriel to work on how to mush together axe-core with Playwright tests
  - Tony and Tania to start working on how to format and store test results

Week 5. Implement 1-2 of Isabela's tests in Galata + Playwright (again, keeping color contrast front of mind)
  - Gabriel to wright (err, write) the Playwright code to turn half of Isabela's tests from human-readable instructions to machine-readable instructions
  - Tony and Gabriel to implement code needed to store tests results
  - Isabela to start planning for next focussed team work cycle

Week 6. Implement rest of Isabela's tests in Galata + Playwright. Make sure results are getting stored in the right place and that test results are shareable via easy mechanism (i.e., GitHub action).
  - Gabriel to finish implementing last of Isabela's tests
  - All hands on deck to help test and debug our integration with Github CI

### Alternatives 

The main thing I want to point out in this section is that the proposal I am making for the next six weeks does not include other crucial aspects of the JupyterLab ecosystem, such as docs, not to mention the even broader Jupyter ecosystem, such as Lumino and RetroLab. This is not to disregard those other parts. It is rather in order to reduce complexity and sharpen focus for this testing phase. This is why I'm calling it "Phase 1."

In order to address this critical shortcoming, I recommend coming back to testing, as a team, a few more times over the course of the grant—i.e., A11y Testing Phase 2, Phase 3, and so on. I wonder if a good approach would be for us to work in cycles, going from testing to fixing to sharing, then back again to testing, fixing, and sharing, and so forth and so on. The reason I think this could be a smart approach is that I believe that when we work on fixing accessibility issues and then on sharing what we have accomplished so far, we will learn things that will inform and improve our approach to testing. There is also a benefit to working on something, putting it aside for a while, and then revisiting it with renewed eyes.

## Addressing concerns

In this section I would like to discuss three concerns about this work. I will point out the strategies we seem to have informally adopted as a team so far, and I will point out where I think there are still open questions.

### Concern 1: Disrupting the developer workflow

Tania has repeatedly voiced a concern that our work should not impede the developer workflow. I think we have managed to address this concern by using the `jupyter/accessibility` repo, following the pattern of the benchmarks repo, so that our envisioned testing infrastructure becomes a value add-on to the developer workflow and not a blocker for any development.

### Concern 2: Maintenance burden

Any software package or system that we build represents a potential maintenance burden for our already small team. Yes, we'll be open-sourcing all of our work, but I fear that at least in the beginning, if anything breaks, it will be on us to fix it. Maintenance becomes an even more crucial issue if our work gains traction (which we want!). We want traction and adoption because that's a measure of success, but that also raises the stakes for when things break. 

I'm not sure what measures or principles we are adopting to address this concern. It seems to me that one thing we should consider, at least in the initial phase, is to avoid putting a software package out there that we intend for developers to use on their own machines. This might be antithetical to the ethos of open source software (I'm not sure); however, I think we will be doing ourselves a solid favor if we clearly communicate that we haven't done the legwork to make sure that the testing code we write can run on a wide range of developer machines. That would free us to focus on making our package run only on specific environments that we control.

Question: what other principles, guidelines, strategies can we follow to help prevent creating a big maintenance burden for our small team?

### Concern 3: Misinterpretation of test results

I have been somewhat squeamish about automated accessibility testing from the beginning because one of the first things I watched after I joined Quansight was a presentation by an engineer who showed how you could [create a website that is literally accessible to no one and yet still get a perfect Lighthouse accessibility score](https://conffab.com/video/building-the-most-inaccessible-site-possible-with-a-perfect-lighthouse-score/).

It's true: developers are responsive to charts and metrics. That is a blessing and a curse, a challenge and an opportunity for the work we do with automated accessibility tests.

I am afraid of building a tool that misdirects human energy. I am still very concerned that our efforts will be misinterpreted.

Unfortunately, having this concern kicking around in my head for months has not yet produced a eureka moment for me. Writing this document has helped give me some clarity about how to message automated accessibility tests (i.e., "avoiding pitfalls" versus "state of accessibility"), but I'm not convinced this will be enough to prevent misinterpretation.

Here's an example scenario that I think we may be setting ourselves up for. Imagine that a JupyterLab developer submits a PR. They know about our suite of accessibility tests and how to run them against their PR. They run the test suite and everything passes—green lights all around (oh what a glorious future!). The developer is thrilled. Their code change has passed our suite of tests. A11y... Check! But then one of us comes along and spots an issue and then for the 100th time we have to explain that, well, um, the tests are not really comprehensive, a green light isn't really a green light etc etc.

Maybe this isn't that significant. Maybe I am making a mountain out of a mole hill. Maybe all of the benefits of automated accessibility tests outweigh the likely misinterpretations. But wherever possible I would like us to try to minimize the kinds of misinterpretations that we can expect, knowing what we know about how software engineers think. 

How do people working in security testing address this issue? (As in, "if your code passes our automated security checks, it doesn't actually mean the code is secure, just that it avoids common security pitfalls.")

Perhaps this is an area in which Isabela could help us come up with good visuals that help to communicate the limitations of our system.

## What's next?

After Phase 1 of a11y testing, I would like to put forward Tony's proposal that we move into a team-coordinated cycle of fixing issues in JupyterLab, beginning with color contrast. Maybe we spend three to six weeks on color contrast. Then spend a couple weeks writing up what we did. Then return to a testing cycle. Then a fixing cycle. Writing up our results. And so on.

**Comment by trallard** [comment]
**☕  I recommend you grab a cup of your favourite brew before reading**

---

> similar challenge of scope is that JupyterLab accessibility issues touch upon so many distinct and disparate areas, not only the JupyterLab app but also JupyterLab docs, extensions, and other critical parts of the ecosystem

This is true and for reference, I created a top-view of what `all these things` in the statement cover within the Jupyter ecosystem... it is a lot and there is for sure no way we make all of these moving pieces together accessible... not us not with the resources we have

![RFD -a11y - core](https://user-images.githubusercontent.com/23552331/164294044-87ecebe7-569e-48d8-929b-839c792ab85f.png)
 
 
 >So what is testing trying to solve? In an ideal world, I think it solves the issue that Tania has mentioned in previous team meetings: giving us a "control room" view of what's broken, what needs to be fixed—letting us know where we are.
 
 I feel we have a very different view of the role and context of "testing" _per se_ (my own personal opinions)
 1. Testing solves **nothing** the purpose of any sort of test is not to fix/remediate/solve issues (this is widely applicable to any area of software development)
 2. What testing is and how it should be treated:
    a. A diagnosis tool - either to reveal the current status of the codebase or to identify potential Root Causes (RA) to issues or unexpected behaviours arisen
    b. A Quality assurance tool: how are we sure that our tool does what we say it is doing? and if it is not, what promises are we breaking? who and how are being affected by these failures? How can I (e.g. a developer/contributor) make sure that what I am adding to the codebase does not break already existing features?
    c. A learning mechanism: what can we learn from the array of passed/failed/inconclusive tests? how can we use this data to learn from and improve our processes and tooling?
    d. An imperfect proxy to identify failing assumptions
    e. Finally, the most common of all - a final switch to prevent bugs from reaching the end-user 
    
Also, note I am not referring to `automated testing` with the above, since `testing != automated testing` - I am talking about a holistic testing approach (or mixed methods approach - automated, manual, etc.)
   
  With those principles' in sight, we **can build** a framework that gives us much valuable information on where we are standing: not only on what is broken and needs fixing but also on what assumptions we are making down the path. 
  It is not a direct metric of `JupyterLab is 85% accessible` because even if we managed to align with all of the WCAG principles and criteria that still does not give us any insight on functional accessibility. 
Maybe  this is a promise not one single tool/IDE/app will make - I have not yet seen anything out there that gives you a hard metric for all levels of accessibility, because I dare to say this is impossible. 
  
  >the most important accessibility issues are impossible to automate
  This statement here makes me feel the broader `testing` is being conflated with `automated testing` though I might be wrong. 
  However, this also opens up another important question - who is responsible for what in the context of Jupyter accessibility, and what should we aim to do to fix it?
  
 So I present exhibit B:
![RFD -a11y - responsibilities](https://user-images.githubusercontent.com/23552331/164296548-a8ed7082-a764-47f0-81fb-453003b63b2c.png)

Many of those concerns about items to be automated do not fall under things we (or the broader Jupyer folks) can solve/nor have responsibility for.  
Since all of these fall in the category of "content generated inside a notebook" the responsibility lies in whoever is generating said content. How can we possibly "test" and fix all notebooks in the wild? We can not - neither we want to.


 So centring back what is **in scope for the grant**?   Enter Exhibit C
 
 
![RFD -a11y - in scope](https://user-images.githubusercontent.com/23552331/164297900-f607fadf-62c9-42bc-a78c-167f351084fd.png)

**JLab** is within our scope - that is, from a software project point of view the sole scope for Phase 1 of the grant. And by sibling association, this will affect other tools that use JLab (i.e. BinderHub, JupyterHub). Anything else is far beyond our promise (I know it sucks, but that is what is for now)


And how can we make a positive, progressive, and more important long-term impact within our "narrow" scope? Enter Exhibit D

![RFD -a11y - workflow (1)](https://user-images.githubusercontent.com/23552331/164299248-d6601e78-e18d-4b92-bbf0-ffb33b0b4815.png)

 This is not one pass, and it's done - but a rather iterative or build-on-iterative where the most valuable piece is not the `unit tests` or even the Playwright scenarios - but the creation of evidence & experience-based practice and documentation for accessibility within JupyterLab (and beyond).
 This workflow has plenty of assumptions but it integrates many of the concerns that @gabalafou  mentioned:
 
 1. How are we disrupting the developer workflow? We have plans to mitigate this -> we might be again making assumptions on this, but the truth is we will not know until we get stuff out and people start using it
 2. Maintenance burden -> all software comes with that, we do want this to become a useful tool and even more so a sustainable one. We have to create participation pathways for folks to ensure sustainability. 
 3. Operate under the assumption (and more often true than not) that users will use your product/tool in unintended ways. There is maybe no way around it but what we can do it set very **explicit** contracts re what our tool does/what it does not/what guarantees we can make (OSI licenses take care of some of this) and document, document, document.
 4. How can we ensure that whatever our tests report on are not misinterpreted - well document document document... make assumptions explicit, clearly word what those numbers mean from a functional and technical point of view. Build supplementary material. That is what the SevDevOps folks do - there is only as much as you can do because accessibility like security is not one single person's job it is everyone's job.

**Comment by trallard** [comment]
Now on the proposal, I do agree with the following:

1. A software package to run a suite of accessibility tests against any app code PR in JupyterLab (with the understanding that `docs` comes later on and is not within the immediate scope of Phase 1)
2. GH actions based report/call to tests 
3. Cron -based automated testing 

Based on the timeline proposed by @gabalafou  - and accounting for the delay in getting the RFD reviewed:

```mermaid

gantt
    title 6 weeks "sprint"
    dateFormat  YYYY-MM-DD
    section Week 1
    RFD - team alignment  :a1, 2022-04-12, 10d
    section Week 2
    Get axe-core tests in a11y repo  :2022-04-21, 7d
    section Week 3
    Setup axe-core tests for storage and sharing :2022-04-28, 7d
    section Week 4
    Add scaffold for Galata + Playwright tests to jupyter/accessibility :2022-05-05, 7d
    section Week 5
    Implement 1-2 of Isabela's tests in Galata + Playwright :2022-05-12, 7d
    section Week 6
    Implement rest of Isabela's tests in Galata + Playwright. :2022-05-19, 7d
      
 ```
 
There are items that are already out of sync, and the original timeline does not account for my PTO, so needs adjusting. 

---

## Decision

That said - with the discussion items **and** the many caveats in the comments, I propose the vote:

👎🏽  - Reject the RFD
👍🏽  - Approve as is
🚀  - Approve the scope, workflows and mitigation plans proposed in https://github.com/Quansight-Labs/jupyter-a11y-mgmt/issues/97#issuecomment-1104334128


@Quansight-Labs/czi-a11y-grant  team vote today by 11:59pm PDT after which the lazy majority will be counted as final decision

**Comment by trallard** [comment]
Note if approved - this will be taken as the source of truth for the upcoming 6 week sprint cc/ @Quansight-Labs/czi-a11y-grant

**Comment by gabalafou** [comment]
I love love love this discussion!

I clicked the 🚀 , let's go! 😸 

The next section is addressed to @trallard's comment.

-----

@trallard, you were spot on in picking up that something that I failed to make explicit in my document is that I narrowed my scope to automated testing. You're right in pointing out that accessibility testing encompasses a much broader scope that includes non-automated, human testing. The question this raises for me is: are there things you think we should be doing on manual testing in year one of the grant? if so, which things?

I also really like how Exhibit B exposed another implicit narrowing of scope in my document. I was only thinking about Jupyter folks (maintainers, developers). I wasn't thinking about content creators and consumers (notebook authors and readers). That is a very important piece of the puzzle when thinking about the accessibility of the entire user experience of reading and writing notebooks in JupyterLab and its associated ecosystem. 

And beyond that, as you show in Exhibit B, the user experience is also shaped by the broader ecosystem (what's happening in browsers, standards, assistive tech, government, etc.) and I think it's important not to lose sight of our position within that ecosystem, as well our push and pull. For example, we may be able to offer a valuable perspective at the W3C.

@trallard, are any of your diagrams in Mural by chance? I would love to be able to comment and ask questions directly on some diagrams. 

As for our differences in the role and context of testing per se, I must have simply misunderstood what I was hearing you say in past meetings. In your comment, the way that you broke automated testing down into 5 parts makes me think we are actually more aligned than I realized (yay!). 

a. Diagnostics ✅ 
b. Quality assurance ✅ 
c. Learning mechanism. I don't think I called this one out explicitly but yes, I totally agree ✅ 
e. Proxy... not sure what this one means
f. Bug prevention ✅ ... although... how is this one different from (b) quality assurance?

The next section is addressed to @isabela-pf's comment.

-----

> I think this requires us documenting the state of tests at different points in time and/or aligned with commits. Is there plans for this, or is this just an idea to consider?

Yes, there are multiple ways that having automated tests can help us "document things along the way."

The first and most basic is just the set of tests that we have. This set should grow over time and that growth will be captured in our git commit history.

Second, the part 3 of my proposal specifically addresses running the tests periodically and capturing the results of those tests. This should allow us to create a time series that helps to document some of our work and progress along the way.

Another way that tests can help to log our progress is if we align them with commits to the JupyterLab repo. As Nick discussed in the JupyterLab accessibility call today, we should be able to plug our automated a11y testing system into the [tools that the maintainers use](https://github.com/jupyterlab/maintainer-tools) to update the change log whenever they cut a new JupyterLab release.

> Naive question, does this refer to the different layouts in JupyterLab? For example, Lab itself, “simple mode,” etc.?

No, I used the term app code mostly to distinguish from documentation source code. But layouts fit within the scope of app code. When running axe-core tests, we'll probably want to test all such layouts.

> Depending what “sharing” includes, I think this could be the least frequent part of the section. If sharing means “demos at JupyterLab or community calls”, for example, I think that could be done in the order you mentioned. If sharing means “documenting and bringing it to other parts of the ecosystem” I think it will be less often.

Yeah, I'm not entirely sure what the form of sharing should take and am totally open on this. I just think it would be a good idea if we build in sessions throughout the grant to work on the part of the grant that has to do with sharing/publishing our results. My mental model of the grant is:

1. Testing
2. Fixing
3. Sharing

As I was working on this six-week proposal, I realized that testing would probably benefit from an on-off rhythm: working on it, then stepping back, then working on it again, then stepping back, etc. Same for fixing. Same for sharing. I'm thinking that rather than save the sharing step for the very end of the grant we could come back to it repeatedly over the course of the grant. 

Tony thinks we should maybe have a sharing or reflection cycle (to be clear, not a six weeks-long cycle!) at the end of each other major cycle of work. We do six weeks of testing, then we work together for a week to collect our findings from that cycle (perhaps in a blog post) and plan the next cycle. Then we could do, for example, a six week fixing cycle and then stop and do another quick retrospective. And so on.

But I should say I'm not opposed to saving the sharing part of the grant for the end. I'm just thinking that it might help with team spirit and public relations if we are periodically celebrating and sharing what we've done with the wider world.
